File: typescript/src/algorithms/leaky-bucket.ts
import { RateLimiter, RateLimiterStats } from './rate-limiter.interface';

interface LeakyBucketData {
  volume: number;
  lastLeak: number;
}

export class LeakyBucket implements RateLimiter {
  private buckets: Map<string, LeakyBucketData> = new Map();
  private capacity: number;
  private leakRate: number;

  constructor(capacity: number, leakRatePerSecond: number) {
    this.capacity = capacity;
    this.leakRate = leakRatePerSecond;
  }

  async allow(key: string): Promise<boolean> {
    const now = Date.now();
    let bucket = this.buckets.get(key);

    if (!bucket) {
      bucket = {
        volume: 0,
        lastLeak: now
      };
      this.buckets.set(key, bucket);
    }

    const timePassed = (now - bucket.lastLeak) / 1000;
    const leaked = timePassed * this.leakRate;
    
    bucket.volume = Math.max(0, bucket.volume - leaked);
    bucket.lastLeak = now;

    if (bucket.volume + 1 <= this.capacity) {
      bucket.volume += 1;
      return true;
    }

    return false;
  }

  async reset(key: string): Promise<void> {
    this.buckets.delete(key);
  }

  async getStats(key: string): Promise<RateLimiterStats> {
    const now = Date.now();
    const bucket = this.buckets.get(key);
    
    if (!bucket) {
      return {
        allowed: true,
        limit: this.capacity,
        remaining: this.capacity,
        resetAt: new Date(now + (this.capacity / this.leakRate) * 1000)
      };
    }

    const timePassed = (now - bucket.lastLeak) / 1000;
    const leaked = timePassed * this.leakRate;
    const currentVolume = Math.max(0, bucket.volume - leaked);
    const remaining = Math.max(0, this.capacity - currentVolume);

    const timeToEmpty = currentVolume / this.leakRate;
    
    return {
      allowed: currentVolume + 1 <= this.capacity,
      limit: this.capacity,
      remaining: Math.floor(remaining),
      resetAt: new Date(now + timeToEmpty * 1000)
    };
  }
}

File: go/algorithms/leaky_bucket.go
package algorithms

import (
	"sync"
	"time"
)

type leakyBucketData struct {
	volume   float64
	lastLeak time.Time
}

type LeakyBucket struct {
	mu       sync.RWMutex
	buckets  map[string]*leakyBucketData
	capacity float64
	leakRate float64
}

func NewLeakyBucket(capacity int, leakRatePerSecond float64) *LeakyBucket {
	return &LeakyBucket{
		buckets:  make(map[string]*leakyBucketData),
		capacity: float64(capacity),
		leakRate: leakRatePerSecond,
	}
}

func (lb *LeakyBucket) Allow(key string) bool {
	lb.mu.Lock()
	defer lb.mu.Unlock()

	now := time.Now()
	b, exists := lb.buckets[key]
	if !exists {
		b = &leakyBucketData{
			volume:   0,
			lastLeak: now,
		}
		lb.buckets[key] = b
	}

	timePassed := now.Sub(b.lastLeak).Seconds()
	leaked := timePassed * lb.leakRate

	b.volume = max(0, b.volume-leaked)
	b.lastLeak = now

	if b.volume+1 <= lb.capacity {
		b.volume++
		return true
	}

	return false
}

func (lb *LeakyBucket) Reset(key string) {
	lb.mu.Lock()
	defer lb.mu.Unlock()
	delete(lb.buckets, key)
}

func (lb *LeakyBucket) GetStats(key string) RateLimiterStats {
	lb.mu.RLock()
	defer lb.mu.RUnlock()

	now := time.Now()
	b, exists := lb.buckets[key]
	if !exists {
		return RateLimiterStats{
			Allowed:   true,
			Limit:     int(lb.capacity),
			Remaining: int(lb.capacity),
			ResetAt:   now.Add(time.Duration(lb.capacity/lb.leakRate) * time.Second),
		}
	}

	timePassed := now.Sub(b.lastLeak).Seconds()
	leaked := timePassed * lb.leakRate
	currentVolume := max(0, b.volume-leaked)
	remaining := max(0, lb.capacity-currentVolume)

	timeToEmpty := currentVolume / lb.leakRate

	return RateLimiterStats{
		Allowed:   currentVolume+1 <= lb.capacity,
		Limit:     int(lb.capacity),
		Remaining: int(remaining),
		ResetAt:   now.Add(time.Duration(timeToEmpty) * time.Second),
	}
}

func max(a, b float64) float64 {
	if a > b {
		return a
	}
	return b
}